src/QsCompiler/Transformations/SearchAndReplace.cs

```diff
@@ -446,6 +446,166 @@ public ExpressionKindTransformation(SyntaxTreeTransformation<TransformationState
         }
     }
 
+    /// <summary>
+    /// A transformation that renames all references to each given qualified name.
+    /// </summary>
+    public class RenameReferences : SyntaxTreeTransformation
```

Comment:
I'm reluctant to use the stateful transformation classes because it requires making the state object public.

Reply:
I am not a big fan of having the TransformationState class be public either, but that's the pattern that the team decided on. It's going to make the transformation code much harder to maintain if every transformation tries to do things differently. Please consider adhering to the pattern.

Reply:
I agree that it would be better to be consistent, but I don't think this pattern is so different from the stateful transformation classes that it would be harder to maintain. It also seems to be the recommended way to access the outer class from within nested classes in C#.

Reply:
I guess the problem we both have with TransformationState being public in the case of this transformation is an issue that is inherent in all stateful transformations that use the pattern. If we make an exception here based on that issue alone, then we should really change the pattern on the whole and do this everywhere. Would you agree with that?

Reply:
It doesn't make sense to me that you would want to make the transformation available publicly, but not the class representing its state. It seems to me that they should either be both public or not (not just by necessity). You can always keep the handles within the transformation call internal to not expose more than you want to, but I don't understand not wanting to have the class exposed?

Reply:
Yes, if nothing in the state class is public, then it is not a big issue, other than having an empty class exposed as part of the public API is a small unnecessary implementation detail. However, the lowest access level I can use for the state class's members is internal. They are currently private the way it is implemented now.

Reply:
Yes, I have had that grievance - I really wish there was a C# access modifier that limits visibility to the class and all containing classes. While it would be nice to keep it fully private, the important piece is that it's not publicly accessible. Similarly with the state class - it's an inner class, so it is not unnecessarily polluting an already large public API (meaning it doesn't make it harder to find other pieces you may want to use, since it would only surface once you are already looking at that particular transformation). 
The maintainance to me is not too much of a concern here - the code needs to be clean, documented and sufficiently understandable for people to maintain it (I think that's the case here). Attempting to keeping everything perfectly the same in my experience tends to break down as the team and project grows, and a lot of little unwritten rules that contributions need to satisfy (or an ever growing set of written rules) only make it harder for people to contribute. That said, I think it is generally good practice and common curtesy to adapt in style and structure when contributing to an existing code base.  

My recommendation is to go with what seems to be a more natural C# pattern here and use a transformation class. If you want to keep it light weight, then you could even choose the dictionary itself as the transformation class, and make GetNewName and RenameUdt private static members that take the dictionary as additional argument. How does that sound?

Reply:
I understand the argument for keeping style consistent within the codebase, but I'm not sure why using a nested state class would be more natural in C#. Microsoft officially [recommends avoiding public nested types](https://docs.microsoft.com/en-us/dotnet/standard/design-guidelines/nested-types). From what I have been able to find, the most common way to use an outer class from a nested class is to pass a reference into the nested class's constructor. 

While there is some overhead in defining a field to hold the reference in each nested class, I think the overhead is comparable to, or lower than, the overhead of creating a new public class and type parameterizing the transformation on it.

Using the dictionary as the state type could work, but is still part of the public API, so any change to how the state is represented (e.g., if later someone decides to replace the dictionary with a custom transformation state class) would be a breaking change.

Reply:
I read through the linked style guide (thanks, that link is helpful). I don't think we are actually violating it and it seems that this case here may be considered to follow the spirit of the listed exceptions. 
I didn't actually see the section where they suggest to store the parent class instance. In our case, it would likely not be one handle, but one handle for each subclass, which seems marginally worse to me. 
Independent on that, keeping subclasses for transformations I believe is a decent choice for the subtransformations given that they are indeed supposed to be used only as part of the parent, and give some structure to what would otherwise just be one gigantic class. 
And a similar argument I think can be made for the transformation state: in a sense, what it is declaring intent and purpose. The idea of the transformation state is that it is a container that is only supposed to be accessed by the transformation itself (and its subclasses) to accumulate temporary information that is needed during execution of the transformation.

Reply:
I'm not opposed to using a separate class to store transformation state, or using nested classes for transformation subtypes (statements, expression, etc.). I would be fine with having a state class that is private to the transformation.

> And a similar argument I think can be made for the transformation state: in a sense, what it is declaring is intent and purpose. The idea of the transformation state is that it is a container that is only supposed to be accessed by the transformation itself (and its subclasses) to accumulate temporary information that is needed during execution of the transformation.

By making it a public class, it communicates the intent that it can be used by anyone, not just the transformation itself. Even if everything in the class is internal, it still communicates the intent that the state could be used by other transformations in the transformations project.

Reply:
I have changed the transformation so it uses a private nested state class now.